metric_learn/pgdm.py

```diff
@@ -0,0 +1,436 @@
+    A0 : (d x d) matrix, optional
+        initial metric, defaults to identity
+        only the main diagonal is taken if `diagonal == True`
+    diagonal : bool, optional
+        if True, a diagonal metric will be learned,
+        i.e., a simple scaling of dimensions
+    diagonal_c : float, optional
+        weight of the dissimilarity constraint for diagonal
+        metric learning
+    verbose : bool, optional
+        if True, prints information while learning
+    """
+    self.max_iter = max_iter
+    self.max_proj = max_proj
+    self.convergence_threshold = convergence_threshold
+    self.A0 = A0
+    self.diagonal = diagonal
+    self.diagonal_c = diagonal_c
+    self.verbose = verbose
+  
+  def fit(self, X, constraints):
+    """Learn the PGDM model.
+    Parameters
+    ----------
+    X : (n x d) data matrix
+        each row corresponds to a single instance
+    constraints : 4-tuple of arrays
+        (a,b,c,d) indices into X, with (a,b) specifying similar and (c,d)
+        dissimilar pairs
+    """
+    constraints = self._process_inputs(X, constraints)
+    if self.diagonal:
+      return self._fit_diag(X, constraints)
+    else:
+      return self._fit_full(X, constraints)
+  
+  def _process_inputs(self, X, constraints):
+    
+    self.X_ = X = check_array(X)
+    
+    # check to make sure that no two constrained vectors are identical
+    a,b,c,d = constraints
+    ident = _vector_norm(X[a] - X[b]) > 1e-9
+    a, b = a[ident], b[ident]
+    ident = _vector_norm(X[c] - X[d]) > 1e-9
+    c, d = c[ident], d[ident]
+    
+    # init metric
+    if self.A0 is None:
+      self.A_ = np.identity(X.shape[1])
+      if not self.diagonal:
+        # Don't know why division by 10... it's in the original code
+        # and seems to affect the overall scale of the learned metric.
+        self.A_ /= 10.0
+    else:
+      self.A_ = check_array(self.A0)
+    
+    return a,b,c,d
+
+  def _fit_full(self, X, constraints):
+    """Learn full metric using PGDM.
+    Parameters
+    ----------
+    X : (n x d) data matrix
+        each row corresponds to a single instance
+    constraints : 4-tuple of arrays
+        (a,b,c,d) indices into X, with (a,b) specifying similar and (c,d)
+        dissimilar pairs
+    """
+    a,b,c,d = constraints
+    num_pos = len(a)
+    num_neg = len(c)
+    num_samples, num_dim = X.shape
+    
+    error1 = error2 = 1e10
+    eps = 0.01        # error-bound of iterative projection on C1 and C2
+    A = self.A_
+    
+    # Create weight vector from similar samples
+    pos_diff = X[a] - X[b]
+    w = np.einsum('ij,ik->jk', pos_diff, pos_diff).ravel()
+    # `w` is the sum of all outer products of the rows in `pos_diff`.
+    # The above `einsum` is equivalent to the much more inefficient:
+    # w = np.apply_along_axis(
+    #         lambda x: np.outer(x,x).ravel(),
+    #         1,
+    #         X[a] - X[b]
+    #     ).sum(axis = 0)
+    t = w.dot(A.ravel() / 100.0)
```

Comment:
Is there an explanation for this division by 100? Combined with the division by 10 in `_prepare_inputs`, there's a lot of magic happening to `A`...